### PR TITLE
#642 - Quick Command Protection

### PR DESCRIPTION
**Pull request by HexelDev**
The point of this Protection is simply detect "suspicious" players in order to kick 'em before the anti bot will be triggered.

I've implemented a manager class that:

- Has an `ExpiringSet` used to store the name for the defined time `QUICK_COMMANDS_DENIED_BEFORE_MILLISECONDS`
  - The name is stored only if the player has the permission `PlayerPermission.QUICK_COMMANDS_PROTECTION`
- Checks if the player is in the `Set`
  - Used to know if the player should be kicked

And yeah, that's it...

## Question:

Should the setting `QUICK_COMMANDS_DENIED_BEFORE_MILLISECONDS` be under the `ProtectionSettings` section (the one used by the anti bot stuffs)  or somewhere else?

## TODO:
- [x] Add kick message
- [x] Check and add players to the set on login
- [x] Check and kick players in set on command execution

### CODE REVIEW

**Comment by ljacqu** [comment]
Not much to say just yet, looks good so far ;) 

The idea is that on join you put the user in the set and then if /login or /register happens it will check if the user is still in the set?

Minor comments:
- The javadoc over the new permission node is the same as the one above.
- If the duration config is `0` or negative will this protection be disabled? I think it does and that's nice, since a lot of our settings already behave like that.
- The new setting is called `denyCommandsBeforeSeconds` but it's actually milliseconds. Please update the name (or drop it and specify it in the comment).
- The `permissionManager` field can be final, too.

**Comment by HexelDev** [comment]
@ljacqu 
> The idea is that on join you put the user in the set and then if /login or /register happens it will check if the user is still in the set?

Exactly, but I don't know if we should check also other commands...

@ljacqu 
> If the duration config is 0 or negative will this protection be disabled? I think it does and that's nice, since a lot of our settings already behave like that.

Should we check it when the class is created or just let the set handle it?

**Comment by ljacqu** [comment]
> Should we check it when the class is created or just let the set handle it?

It's probably fine just like that, since it's pretty clear behavior from the set imho :)

**Comment by HexelDev** [comment]
I'm now writing the test class, and I'll generate the mesages etc...

**Comment by HexelDev** [comment]
This should be ready, what do you think about it? 

@ljacqu @sgdc3

**Comment by sgdc3** [comment]
Good for me ;)

File: src/main/java/fr/xephi/authme/listener/PlayerListener.java [hunk 1]
```diff
@@ -207,6 +215,11 @@ public void onPlayerJoin(PlayerJoinEvent event) {
         if (!PlayerListener19Spigot.isPlayerSpawnLocationEventCalled()) {
             teleportationService.teleportOnJoin(player);
         }
+
+        if (quickCommandsProtectionManager.shouldSaveLogin(player)) {
+            quickCommandsProtectionManager.setLogin(player.getName());
+        }
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/listener/PlayerListener.java]: I would personally merge both methods into something like `processLogin(player)` (or saveLogin or whatever) and the cmd protection manager itself takes care of checking the permission node.

Comment (HexelDev) [reply src/main/java/fr/xephi/authme/listener/PlayerListener.java]: Perfect, editing ;)

**Pull request merged**